### PR TITLE
Always use a dedicated artifact for warmup function

### PR DESCRIPTION
**Pull request by richarddd**
Added zip functionality so that this plugin will always have a dedicated artifact.

Fixes: https://github.com/FidelLimited/serverless-plugin-warmup/issues/25

This also solves issues when using a single artifact for all other functions when using for example `java` or `c#`

### CODE REVIEW

**Comment by marcfielding1** [comment]
I'm not sure how active this repo is, I'm waiting for feedback on [#24](https://github.com/FidelLimited/serverless-plugin-warmup/issues/24) for nearly two months - I was actually either going to fork this repo or create a new plugin this weekend if I hadn't back, @goncaloneves @rschweizer is this repo still being maintained?

File: src/index.js [hunk 1]
```diff
@@ -85,6 +102,13 @@ class WarmUP {
    * */
   afterDeployFunctions () {
     this.configPlugin()
+    if (this.warmup.cleanFolder) {
```
Comment (juanjoDiaz) [inline src/index.js]: You can do it without duplicating code (DRY).
```
    let promise = Promise.resolve();  

    if (this.warmup.cleanFolder) {
      promise = promise.then(() => fs.removeAsync(this.pathArchive))
    }

    if (this.warmup.prewarm) {
      promise = promise.then(this.warmUpFunctions))
    }

    return promise;
```

File: src/index.js [hunk 2]
```diff
@@ -49,7 +66,7 @@ class WarmUP {
    * */
   afterPackageInitialize () {
     // See https://github.com/serverless/serverless/issues/2631
-    this.options.stage  = this.options.stage
+    this.options.stage = this.options.stage
```
Comment (juanjoDiaz) [inline src/index.js]: Nice that you fix this style issues!

File: src/index.js [hunk 3]
```diff
@@ -39,6 +43,19 @@ class WarmUP {
     }
   }
 
+  archiveDirectory (src, dest, format = 'zip', { archiverOptions = {}, directoryDestPath = false } = {}) {
```
Comment (juanjoDiaz) [inline src/index.js]: Why having some normal properties and then some properties destructure from an object?

I'd flatten it completely `archiveDirectory (src, dest, format = 'zip', archiverOptions = {}, directoryDestPath = false)`

**Comment by juanjoDiaz** [comment]
Maybe I'm missing something but just running `serverless package` in any of my projects I get a zip file containing all my lambdas (including `_warmup`) and I also get a zip called `warmUpPlugin` containing this plugin.

So I'm trying to understand what's this PR providing apart from that?

**Comment by juanjoDiaz** [comment]
Thanks, @richarddd! That clarifies the use case.

As I have commented in #25, would it be a solution to this problem to simply apply the `package: artifact:` at the function level instead of the service level?

So, instead of
```
package:
  artifacts: path/to/my-artifact.zip

functions:
  myFunction:
    ...
```
use
```
functions:
  myFunction:
    ...
    package:
      artifacts: path/to/my-artifact.zip
```

That way, serverless still packages the lambdas individually and correctly generates the package for the warmup plugin.

**Comment by richarddd** [comment]
@juanjoDiaz I have to try this. Does the framework just upload one artifact for all (other) functions? Or is it copied? I do not wish to have multiple artifacts for all other lambdas except for this plugin

**Comment by juanjoDiaz** [comment]
Hmmm, I was assuming that you have a single function pointing to your artifact. 
If you have multiple functions pointing to the same artifact I guess that it might get verbose to repeat it.

The real issue here is simply that when you have a top level `package: artifact`, Serverless doesn't package individual functions even if they are set as `individually: true`. So it's not really a problem of this plugin but a problem in Serverless package mechanism.

I just did a PR to Serverless fixing that, let's hope that it gets merged.

I think that your workaround is good. But I'd much rather fix the root cause than just work around  it 🙂

**Comment by richarddd** [comment]
Thank you for your comment. Fully agree!

I dont know if you noticied @juanjoDiaz  but i really want to add this `ClientContext` to params as well:
```
const params = {
      // base64 encoded: {"custom":{"warmupEvent":"1"}}
      ClientContext: "eyJjdXN0b20iOnsid2FybXVwRXZlbnQiOiIxIn19",
```

**Comment by juanjoDiaz** [comment]
Hi,

https://github.com/serverless/serverless/pull/5364 just got merge which essentially fix the packaging issues.

Now (on next Serverless release) you can do:
```
service: test

provider:
  name: aws
  runtime: nodejs8.10

package:
  artifact: my-app.zip

functions:
  normalFunc:
    handler: func/handler.func
    events:
      - http:
          path: func
          method: post
  individualFunc:
    handler: func/handler.func
    events:
      - http:
          path: func2
          method: post
    package:
       individually: true
```
So running serverless package with the above config will use `my-app.zip` as artifact for all functions but will create a zip for the ones set as `package: individually: true` like `func/handler.func` (the func folder needs to exist of course).

Since serverless-plugin-warmup is set as `package: individually: true`, it will always get its own zip.
So no need for custom zipping.


Regarding the client context thingy, it seems like a pretty unrelated change to packaging. Could elaborate a bit on what's the use case?

**Comment by richarddd** [comment]
Awesome!

True! I should do another PR regarding that!

**Pull request closed without merging**